File: src/commands/env/status.rs
use colored::*;
use std::env;
use std::fs;

use crate::commands::setup::utils as setup_utils;
use crate::vault;

use super::utils as env_utils;

pub fn run() {
    if let Err(e) = run_inner() {
        eprintln!("{} {}", "Status failed:".red().bold(), e);
        crate::terminal::cleanup_and_exit(1);
    }
}

fn run_inner() -> Result<(), String> {
    let cwd = env::current_dir()
        .map_err(|e| format!("Cannot read current directory: {}", e))?;

    // Derive project ID
    let (project_id, used_fallback) = env_utils::derive_project_id(&cwd)?;
    if used_fallback {
        println!(
            "  {} {}",
            "!".yellow().bold(),
            format!(
                "No git remote found. Using folder name '{}' as project identifier.",
                project_id
            ).dimmed()
        );
    }
    println!(
        "  {} {}",
        "Project:".dimmed(),
        project_id.cyan().bold()
    );

    // Resolve vault and fetch latest (best-effort)
    let vault_dir = vault::vault_dir();
    let vault_dir_str = vault_dir.display().to_string();

    let sp = setup_utils::spinner("Fetching latest...");
    let fetch_failed = setup_utils::run_git(&vault_dir_str, &["fetch"]).is_err();
    sp.finish_and_clear();
    if fetch_failed {
        println!(
            "  {} {}",
            "!".yellow().bold(),
            "Could not fetch latest — showing local vault state.".dimmed()
        );
    }

    // Check vault for this project
    let vault_env_dir = env_utils::envs_dir(&vault_dir, &project_id);
    let vault_files: Vec<String> = if vault_env_dir.exists() {
        fs::read_dir(&vault_env_dir)
            .map_err(|e| format!("Failed to read vault envs: {}", e))?
            .filter_map(|e| e.ok())
            .map(|e| e.file_name().to_string_lossy().to_string())
            .filter(|name| name.starts_with(".env"))
            .collect()
    } else {
        Vec::new()
    };

    // Check local .env files
    let local_files: Vec<String> = env_utils::discover_env_files(&cwd)
        .iter()
        .filter_map(|p| p.file_name())
        .map(|n| n.to_string_lossy().to_string())
        .collect();

    // Compare
    let in_both: Vec<&String> = local_files.iter().filter(|f| vault_files.contains(f)).collect();
    let local_only: Vec<&String> = local_files.iter().filter(|f| !vault_files.contains(f)).collect();
    let vault_only: Vec<&String> = vault_files.iter().filter(|f| !local_files.contains(f)).collect();

    println!();

    if vault_files.is_empty() && local_files.is_empty() {
        println!(
            "  {} No .env files found locally or in the vault.",
            "·".dimmed()
        );
        return Ok(());
    }

    if !in_both.is_empty() {
        println!("  {} Synced (in vault and local):", "■".green());
        for f in &in_both {
            println!("    {}", f.cyan());
        }
    }

    if !local_only.is_empty() {
        println!("  {} Local only (not pushed):", "■".yellow());
        for f in &local_only {
            println!("    {}", f.yellow());
        }
    }

    if !vault_only.is_empty() {
        println!("  {} Vault only (not pulled):", "■".blue());
        for f in &vault_only {
            println!("    {}", f.blue());
        }
    }

    println!();
    Ok(())
}

File: ts/src/commands/env/status.ts
// `keypick env status` — compare local .env files with vault copies.
// Ported from rust/src/commands/env/status.rs.

import { existsSync, readdirSync } from "node:fs";
import path from "node:path";
import process from "node:process";
import chalk from "chalk";
import * as vault from "../../lib/vault.ts";
import * as terminal from "../../lib/terminal.ts";
import * as setupUtils from "../setup/utils.ts";
import * as envUtils from "./utils.ts";

export async function run(): Promise<void> {
  try {
    await runInner();
  } catch (e) {
    console.error(chalk.red.bold("Status failed:"), (e as Error).message);
    terminal.cleanupAndExit(1);
  }
}

async function runInner(): Promise<void> {
  const cwd = process.cwd();
  const { id: projectId, usedFallback } = envUtils.deriveProjectId(cwd);
  if (usedFallback) {
    console.log(
      `  ${chalk.yellow.bold("!")} ${chalk.dim(
        `No git remote found. Using folder name '${projectId}' as project identifier.`,
      )}`,
    );
  }
  console.log(`  ${chalk.dim("Project:")} ${chalk.cyan.bold(projectId)}`);

  const vaultDir = await vault.vaultDir();

  const sp = setupUtils.spinner("Fetching latest...");
  let fetchFailed = false;
  try {
    setupUtils.runGit(vaultDir, ["fetch"]);
  } catch {
    fetchFailed = true;
  }
  sp.stop();
  if (fetchFailed) {
    console.log(
      `  ${chalk.yellow.bold("!")} ${chalk.dim("Could not fetch latest — showing local vault state.")}`,
    );
  }

  const vaultEnvDir = envUtils.envsDir(vaultDir, projectId);
  let vaultFiles: string[] = [];
  if (existsSync(vaultEnvDir)) {
    try {
      vaultFiles = readdirSync(vaultEnvDir).filter((n) => n.startsWith(".env"));
    } catch (e) {
      throw new Error(`Failed to read vault envs: ${(e as Error).message}`);
    }
  }

  const localFiles: string[] = envUtils.discoverEnvFiles(cwd).map((p) => path.basename(p));

  const vaultSet = new Set(vaultFiles);
  const localSet = new Set(localFiles);
  const inBoth = localFiles.filter((f) => vaultSet.has(f));
  const localOnly = localFiles.filter((f) => !vaultSet.has(f));
  const vaultOnly = vaultFiles.filter((f) => !localSet.has(f));

  console.log();

  if (vaultFiles.length === 0 && localFiles.length === 0) {
    console.log(`  ${chalk.dim("·")} No .env files found locally or in the vault.`);
    return;
  }

  if (inBoth.length > 0) {
    console.log(`  ${chalk.green("■")} Synced (in vault and local):`);
    for (const f of inBoth) console.log(`    ${chalk.cyan(f)}`);
  }
  if (localOnly.length > 0) {
    console.log(`  ${chalk.yellow("■")} Local only (not pushed):`);
    for (const f of localOnly) console.log(`    ${chalk.yellow(f)}`);
  }
  if (vaultOnly.length > 0) {
    console.log(`  ${chalk.blue("■")} Vault only (not pulled):`);
    for (const f of vaultOnly) console.log(`    ${chalk.blue(f)}`);
  }

  console.log();
}
